libraries/RC_Channel/RC_Channel.cpp

```diff
@@ -407,32 +412,38 @@ void RC_Channel::read_mode_switch()
     else if (pulsewidth < 1750) position = 4;
     else position = 5;
 
-    if (mode_switch_state.last_position == position) {
-        // nothing to do
-        return;
-    }
-
-    const uint32_t tnow_ms = AP_HAL::millis();
-    if (position != mode_switch_state.debounced_position) {
-        mode_switch_state.debounced_position = position;
-        // store time that switch last moved
-        mode_switch_state.last_edge_time_ms = tnow_ms;
-        return;
-    }
-
-    if (tnow_ms - mode_switch_state.last_edge_time_ms < MODE_SWITCH_DEBOUNCE_TIME_MS) {
-        // still in debounce
+    if (!debounce_completed(position)) {
         return;
     }
 
     // set flight mode and simple mode setting
     mode_switch_changed(position);
+}
+
+bool RC_Channel::debounce_completed(int8_t position) 
+{
+    // switch change not detected
+    if (switch_state.current_position == position) {
+        // reset debouncing
+         switch_state.debounce_position= position;
+    } 
+    // switch change detected
+    else {
+        const uint32_t tnow_ms = AP_HAL::millis();
+
+        // position not established yet
+        if (switch_state.debounce_position != position) {
+            switch_state.debounce_position = position;
+            switch_state.last_edge_time_ms = tnow_ms;
+        } 
+        // position estabilished; debounce completed
+        else if (tnow_ms - switch_state.last_edge_time_ms >= (uint16_t)MAX(debounce_time_ms, SWITCH_DEBOUNCE_TIME_MIN_MS)) {
```

Comment:
Same.

Reply:
since it's a param that has the possibility to go crazy, we should enforce a max value as well. Perhaps 2000?

Reply:
I'm surprised to see the existing driver timer work this way, it can easily be re-written to not need a debounce_position by simply setting last_edge_time_ms =0

```
bool debounce_done = false;
if (switch_state.current_position == position) {
  switch_state.last_edge_time_ms = 0;
  debounce_done = true;
} else if if (switch_state.last_edge_time_ms == 0) {
  switch_state.last_edge_time_ms = tnow_ms;
} else if (tnow_ms - switch_state.last_edge_time_ms >= timeout) {
  switch_state.current_position = position;
  switch_state.last_edge_time_ms = 0;
  debounce_done = true;
}
return debounce_done;
```

Reply:
don't forget to reset the timer in reset_mode_switch()

Reply:
@magicrub I don't think we shall limit max value of the debounce timeout. The parameter type is Int16 so max is about 34 seconds. This is pretty large number but still not something crazy.

>>it can easily be re-written to not need a debounce_position by simply setting last_edge_time_ms =0

Note that the whole debounce_completed() method must return true, ONLY if the switch was really **changed** and debounced. Your variant does always return true, thus it will constantly raise a switch event.

Reply:
Code style fixed